src/lightning/fabric/utilities/load.py

```diff
@@ -227,3 +235,76 @@ def _move_state_into(
             destination[key].load_state_dict(state)
         else:
             destination[key] = state
+
+
+def _load_distributed_checkpoint(checkpoint_folder: Path) -> Dict[str, Any]:
+    """Loads a sharded checkpoint saved with the `torch.distributed.checkpoint` into a full state dict.
+
+    The current implementation assumes that the entire checkpoint fits in CPU memory.
+
+    """
+    if not _TORCH_GREATER_EQUAL_2_1:
+        raise ImportError("Processing distributed checkpoints requires PyTorch >= 2.1.")
+
+    from torch.distributed.checkpoint import FileSystemReader
+    from torch.distributed.checkpoint.metadata import BytesStorageMetadata, TensorStorageMetadata
+
+    if _TORCH_GREATER_EQUAL_2_2:
+        from torch.distributed.checkpoint import load
+    else:
+        from torch.distributed.checkpoint import load_state_dict as load  # deprecated
+
+    reader = FileSystemReader(checkpoint_folder)
+    metadata = reader.read_metadata()
+
+    # TODO: Add sequential save to avoid storing the entire checkpoint in memory
+    checkpoint: Dict[str, Any] = {}
+    for tensor_name, sd_metadata in metadata.state_dict_metadata.items():
+        if isinstance(sd_metadata, BytesStorageMetadata):
+            checkpoint[tensor_name] = "<bytes_io>"
```

Comment:
For my knowledge, what does this represent?

Reply:
They serialize tensors in a special format, and if it's not a tensor, it's getting serialized as raw bytes. For example, a list of float values like in the optimizer state dict is not a tensor object. By assigning this special string to the state, we tell the state dict loader to deserialize the object from bytes. I think it's just using the regular pickle module. Found this buried deep in the PyTorch code since there is no documentation. This is likely going to break in future versions but this is life 🤷